src/Control/Distributed/Process/Internal/Primitives.hs

```diff
@@ -982,7 +983,10 @@ expectTimeout n = receiveTimeout n [match return]
 spawnAsync :: NodeId -> Closure (Process ()) -> Process SpawnRef
 spawnAsync nid proc = do
   spawnRef <- getSpawnRef
-  sendCtrlMsg (Just nid) $ Spawn proc spawnRef
+  node     <- getSelfNode
+  if nid == node
+    then sendCtrlMsg Nothing $! Spawn (DS.force proc) spawnRef
```

Comment:
> Can't we push this logic to sendCtrlMsg?

How?

> In any case NFData is overkill here: Spawn is strict in its first argument, and Closure doesn't normally embed thunks.

That is not true in our projects. I found at least one case where deep evaluation of the closure produced an error, and it was an expensive error to find.